ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py

```diff
@@ -0,0 +1,263 @@
+        at minimum. 
+        """
+        param_keys = ["strength", "gamma", "demo_path"]
+        super().check_config(config_dict, param_keys)
+
+    def update(self, update_buffer: Buffer, n_sequences: int) -> Dict[str, float]:
+        """
+        Updates model using buffer.
+        :param update_buffer: The policy buffer containing the trajectories for the current policy.
+        :param n_sequences: The total (demo + environment) number of sequences used in each mini batch.
+        :return: The loss of the update.
+        """
+        batch_losses = []
+        n_sequences = max(n_sequences // 2, 1)  # Divide by 2 since we have two buffers
+        possible_demo_batches = (
+            len(self.demonstration_buffer.update_buffer["actions"]) // n_sequences
+        )
+        possible_policy_batches = len(update_buffer["actions"]) // n_sequences
+        possible_batches = min(possible_policy_batches, possible_demo_batches)
+
+        if self.print_debug:
+            kl_loss = []
+            policy_estimate = []
+            expert_estimate = []
+            z_log_sigma_sq = []
+            z_mean_expert = []
+            z_mean_policy = []
+
+        n_epoch = self.num_epoch
+        for _epoch in range(n_epoch):
+            self.demonstration_buffer.update_buffer.shuffle()
+            update_buffer.shuffle()
+            if self.max_batches == 0:
+                num_batches = possible_batches
+            else:
+                num_batches = min(possible_batches, self.max_batches)
+            for i in range(num_batches):
+                demo_update_buffer = self.demonstration_buffer.update_buffer
+                policy_update_buffer = update_buffer
+                start = i * n_sequences
+                end = (i + 1) * n_sequences
+                mini_batch_demo = demo_update_buffer.make_mini_batch(start, end)
+                mini_batch_policy = policy_update_buffer.make_mini_batch(start, end)
+                run_out = self._update_batch(mini_batch_demo, mini_batch_policy)
+                loss = run_out["gail_loss"]
+
+                if self.print_debug:
+                    policy_estimate.append(run_out["policy_estimate"])
+                    expert_estimate.append(run_out["expert_estimate"])
+                    if self.model.use_vail:
+                        kl_loss.append(run_out["kl_loss"])
+                        z_log_sigma_sq.append(run_out["z_log_sigma_sq"])
+                        z_mean_policy.append(run_out["z_mean_policy"])
+                        z_mean_expert.append(run_out["z_mean_expert"])
+
+                batch_losses.append(loss)
+        self.has_updated = True
+
+        if self.print_debug:
+            print_list = ["n_epoch", "beta", "policy_estimate", "expert_estimate"]
+            print_vals = [
+                n_epoch,
+                self.policy.sess.run(self.model.beta),
```

Comment:
Does this have any side effects? Should it be outside the `if self.print_debug`?

Reply:
removed self.print_debug entirely. Am now using the logging.debug() call.